examples/generative/ddpm.py

```diff
@@ -0,0 +1,682 @@
+    )
+    temb = layers.Dense(units, kernel_initializer=kernel_init(1.0))(temb)
+    return temb
+
+
+def build_network():
+    image_input = layers.Input(
+        shape=(img_size, img_size, img_channels), name="image_input"
+    )
+    time_input = keras.Input(shape=(), dtype=tf.int64, name="time_input")
+
+    x = layers.Conv2D(
+        first_conv_channels,
+        kernel_size=(3, 3),
+        padding="same",
+        kernel_initializer=kernel_init(1.0),
+    )(image_input)
+
+    temb = layers.Lambda(TimeEmbedding, arguments={"dim": first_conv_channels * 4})(
+        time_input
+    )
+    temb = TimeMLP(temb, units=first_conv_channels * 4)
+
+    skips = [x]
+
+    # DownBlock
+    for i in range(len(widths)):
+        for _ in range(num_res_blocks):
+            x = ResidualBlock(widths[i], groups=norm_groups)([x, temb])
+            if has_attention[i]:
+                x = AttentionBlock(widths[i], groups=norm_groups)(x)
+            skips.append(x)
+
+        if widths[i] != widths[-1]:
+            x = DownSample(widths[i])(x)
+            skips.append(x)
+
+    # MiddleBlock
+    x = ResidualBlock(widths[-1], groups=norm_groups)([x, temb])
+    x = AttentionBlock(widths[-1], groups=norm_groups)(x)
+    x = ResidualBlock(widths[-1], groups=norm_groups)([x, temb])
+
+    # UpBlock
+    for i in reversed(range(len(widths))):
+        for _ in range(num_res_blocks + 1):
+            x = layers.Concatenate(axis=-1)([x, skips.pop()])
+            x = ResidualBlock(widths[i], groups=norm_groups)([x, temb])
+            if has_attention[i]:
+                x = AttentionBlock(widths[i], groups=norm_groups)(x)
+
+        if i != 0:
+            x = UpSample(widths[i])(x)
+
+    # End block
+    x = layers.GroupNormalization(groups=norm_groups)(x)
+    x = act_fn(x)
+    x = layers.Conv2D(3, (3, 3), padding="same", kernel_initializer=kernel_init(0.0))(x)
+    return keras.Model([image_input, time_input], x, name="unet")
+
+
+"""
+## Training
+
+We will follow the same setup for training the diffusion model as described
+in the paper. We will use `Adam` optimizer with a learning rate of `2e-4`.
+We will use EMA on model parameters with a decay factor of 0.999. We will
+treat our model as noise prediction network i.e. at every training step, we
+will input a batch of images and corresponding time steps to our UNet,
+and the network will output the noise as predictions.
+
+The only difference is that we will not be using the Kernel Inception Distance (KID)
+or Frechet Inception Distance (FID) for evaluating the quality of generated
+samples during training. This is because both these metrics are compute heavy
+and are skipped for the brevity of implementation.
+
+**Note: ** We are using mean squared error as the loss function which is aligned with
+the paper, and theoretically makes sense. In practice, though, it is also common to
+use mean absolute error or Huber loss as the loss function.
+"""
+
+
+class DiffusionModel(keras.Model):
```

Comment:
Would it be useful to implement a `call()` method?

Reply:
Any specific reason for that?